tyrian/js/src/main/scala/tyrian/runtime/TyrianRuntime.scala

```diff
@@ -1,134 +1,97 @@
-  @SuppressWarnings(Array("scalafix:DisableSyntax.var"))
-  private var currentSubscriptions: List[(String, F[Unit])] = Nil
-  // This is a queue of new subs waiting to be run for the first time.
-  // In the event that two events happen at once, you can't assume that
-  // you would have run all the subs between events.
-  @SuppressWarnings(Array("scalafix:DisableSyntax.var"))
-  private var aboutToRunSubscriptions: Set[String] = Set.empty
-
-  // Update
-
-  private def onMsg(msg: Msg): Unit =
-    val res: F[Unit] =
-      for {
-        currentModel <- model.get
-        (updatedState, cmd) = update(currentModel.model)(msg)
-        _ <- completeUpdate(cmd, updatedState)
-      } yield ()
-
-    dispatcher.unsafeRunAndForget(res)
-
-  private def completeUpdate(cmd: Cmd[F, Msg], updatedState: Model): F[Unit] =
-    for {
-      _           <- model.set(ModelHolder(updatedState, true))
-      sideEffects <- gatherSideEffects(cmd, subscriptions(updatedState))
-      _           <- Stream.emits(sideEffects).foreach(msg => channel.send(msg).void).compile.drain
-    } yield ()
+object TyrianRuntime:
+
+  def apply[F[_], Model, Msg](
+      node: Element,
+      initModel: Model,
+      initCmd: Cmd[F, Msg],
+      update: Model => Msg => (Model, Cmd[F, Msg]),
+      view: Model => Html[Msg],
+      subscriptions: Model => Sub[F, Msg]
+  )(using F: Async[F]): F[Nothing] = Dispatcher.sequential[F].use { dispatcher =>
+    (F.ref(ModelHolder(initModel, true)), AtomicCell[F].of(List.empty[(String, F[Unit])]), Channel.unbounded[F, Msg])
+      .flatMapN { (model, currentSubs, msgs) =>
+
+        def runCmd(cmd: Cmd[F, Msg]): Stream[F, Nothing] =
+          Stream
+            .emits(CmdHelper.cmdToTaskList(cmd))
+            .parEvalMapUnorderedUnbounded(_.handleError(_ => None))
+            .unNone
+            .foreach(msgs.send(_).void)
+
+        def runSub(sub: Sub[F, Msg]): F[Unit] =
+          currentSubs.evalUpdate { oldSubs =>
+            val allSubs                 = SubHelper.flatten(sub)
+            val (stillAlive, discarded) = SubHelper.aliveAndDead(allSubs, oldSubs)
+
+            val newSubs = SubHelper
+              .findNewSubs(allSubs, stillAlive.map(_._1), Nil)
+              .traverse(SubHelper.runObserve(_) { result =>
+                dispatcher.unsafeRunAndForget(
+                  OptionT(F.fromEither(result)).foreachF(msgs.send(_).void)
+                )
+              })
+
+            discarded.traverse_(_.start) *> newSubs.map(_ ++ stillAlive)
+          }
+        // end runSub
+
+        val msgLoop = msgs.stream.evalMap { msg =>
+          model
+            .modify { case ModelHolder(oldModel, _) =>
+              val (newModel, cmd) = update(oldModel)(msg)
+              val sub             = subscriptions(newModel)
+              (ModelHolder(newModel, true), (cmd, sub))
+            }
+            .flatMap { (cmd, sub) =>
+              runSub(sub).as(runCmd(cmd))
+            }
+        }.parJoinUnbounded
```

Comment:
We're ok to do this in parallel because JS is single threaded?

Reply:
Hm, actually I think this is necessary for correct semantics. Because otherwise a side-effect like an `IO.sleep(...)` or some network request could block processing of other cmds/subs.

By running in parallel, it's enabling other work to be done while some tasks are asynchronous suspended.

Reply:
Yes you're right, I see what's going on now - thanks. Just thinking about this out loud:

When _I_ think about the update loop I tend to think of the loop - quite wrongly - that the messages and effects are strictly ordered, and they are emitted in order, but of course that's wrong and they may take variable lengths of time to arrive if - as you say - they contained a callback or a sleep or something.

In practice the fact that messages and their results don't resolve in their original order is a) not a real problem* and b) impractical because it would mean blocking on everything.

(*The way you end up programming things in Tyrian / Elm means that either you're waiting for your own messages to trigger an onward journey, thus order is enforced by design, or order doesn't matter and you just want everything to happen as quickly as possible.)

Reply:
The * is a nice observation :)